### DISCUSSION TITLE
Can Quarto handle external assets (e.g. for HTML widgets) delivered over CDN?

### DISCUSSION THREAD

**Discussion by walkerke** [quarto-dev/quarto-cli#10325] [Q&A] [answered]
### Description

I'm developing a package called __mapgl__, installable at `remotes::install_github("walkerke/mapgl")`.  For licensing reasons, I'm bringing in the Mapbox GL JS JavaScript and CSS assets in via CDN rather than bundling them within the package.  However, this gives Quarto problems: 

````qmd
---
title: "quarto_test"
format: html
editor: visual
---

```{r}
library(mapgl)

mapboxgl()
```
````

Gives: 

```
Error in FUN(X[[i]], ...) : 
  Dependency mapbox-gl-js 3.4.0 is not disk-based
Calls: .main ... dependencies_from_render -> html_dependencies_as_string -> lapply -> FUN
Execution halted
```

I know this isn't a problem with my package _per se_, as disk-based assets render fine in Quarto.  However - I do notice that R Markdown does not have the same problem ([see my package documentation here, which I built in R Markdown for that reason](https://walker-data.com/mapgl/)).  

Is there any workaround you know of to be able to pull in external assets via CDN?  Or, does this merit filing an issue?

**Comment by mcanouil** [comment]
The error does not come from Quarto. It's from `htmltools::copyDependencyToDir`.

This function is used inside Quarto using the default, which is `mustWork = TRUE`.
>  If `TRUE` and `dependency` does not point to a directory on disk (but rather a URL location), an error is raised. If `FALSE` then non-disk dependencies are returned without modification.

- https://github.com/quarto-dev/quarto-cli/blob/7de42537a566a0aff7c1396e39c70be598f746f3/src/resources/rmd/execute.R#L498-L513

So right now, Quarto does not allow dependencies to be URL.

It's possible to change this behaviour, but it's not simply changing `mustWork = TRUE` to `mustWork = FALSE`.

CC. @cderv

**Reply by cderv** [reply]
We do use `mustWork = FALSE` in **rmarkdown** context
https://github.com/cderv/rmarkdown/blob/d0088950258f3835a9de9bafa0cb470e4040e3a9/R/html_dependencies.R#L316-L324

> So right now, Quarto does not allow dependencies to be URL.

I don't know if this is by design or an oversight. I need to try some things with the feature we have. Usually we expect htmlDependency to provide on-disk version. 

Thanks for the report.

**Reply by cderv** [reply, marked as answer]
I have opened an issue to track 
-  #10339

**Comment by walkerke** [comment]
Thank you @cderv and @mcanouil!  

Just a thought as well - I'm noticing that there are issues rendering HTML widgets over in Positron (https://github.com/posit-dev/positron/issues/2023) and I wonder if there is some relationship there (based on the discussion in https://github.com/posit-dev/positron/issues/1742).  

Perhaps of interest to @juliasilge @jmcphers - I have the exact same behavior in Positron (Mapbox won't render, but MapLibre does, and I bundle the MapLibre assets).

**Reply by jmcphers** [reply]
Positron's web views for interactive content (inherited from Code OSS) have pretty restrictive security settings and will only load content from allowlisted domains. If you open the Developer Tools (Toggle Developer Tools on the Command Palette), then switch to the Network tab in the tools, then try to render your widget, you can see there if requests to the CDN are being blocked (my guess is they are).

**Reply by mcanouil** [reply]
- https://github.com/quarto-dev/quarto-cli/issues/10339

**Reply by walkerke** [reply]
@jmcphers just as a heads up - this issue in Positron is now solved with the latest release!  Thank you so much!

**Reply by jmcphers** [reply]
great, glad to hear it!
